lib, treewide: introduce `repoRevToName` and use it to cleanup most `fetch*` functions

oxij opened the pull request:
This a first half of #49862. Basically, this is a noop cleanup that keeps all the `fetch*` names as they are now, while introducing the new `config.fetchedSourceNameDefault` option.

# Changes

This patch adds `lib.repoRevToName` function that generalizes away most of the code used for derivation name generation by `fetch*` functions (`fetchzip`, `fetchFromGitHub`, etc, except those which are delayed until latter commits for mass-rebuild reasons).

It's first argument controls how the resulting name will look (see below).

Since `lib` has no equivalent of Nixpkgs' `config`, this patch adds `config.fetchedSourceNameDefault` option to Nixpkgs and then re-exposes `lib.repoRevToName config.fetchedSourceNameDefault` expression as `pkgs.repoRevToNameMaybe` which is then used in `fetch*` derivations.

The result is that different values of `config.fetchedSourceNameDefault` now control how the `src` derivations produced by `fetch*` functions are to be named, e.g.:

- `fetchedSourceNameDefault = "source"` (the default):

        $ nix-instantiate -A fuse.src
        /nix/store/<hash>-source.drv

- `fetchedSourceNameDefault = "versioned"`:

        $ nix-instantiate -A fuse.src
        /nix/store/<hash>-libfuse-2.9.9-source.drv

- `fetchedSourceNameDefault = "full"`:

        $ nix-instantiate -A fuse.src
        /nix/store/<hash>-libfuse-2.9.9-github-source.drv

See documentation of `config.fetchedSourceNameDefault` for more info.

AndersonTorres commented on lib/sources.nix:
```diff
@@ -254,6 +254,64 @@ let
+  urlToName = url_:
+    let
+      inherit (builtins) stringLength;
+      url = toString url_;
+      base = baseNameOf (lib.removeSuffix "/" (lib.last (lib.splitString ":" url)));
```
```suggestion
      base = baseNameOf (lib.removeSuffix "/" (lib.last (lib.splitString ":" (toString url))));
```

AndersonTorres commented on lib/sources.nix:
```diff
@@ -254,6 +254,64 @@ let
 
+  # urlToName : (URL | Path | String) -> String
+  #
+  # Transform a URL (or path, or string) into a clean package name.
```
Maybe I can be wrong, but the general term is URI.

oxij replied:
As a thing more general than a URL? Yes. But here it is literally "a URL, or a path, or a string (a package name, usually)".

AndersonTorres commented on lib/sources.nix:
```diff
@@ -254,6 +254,64 @@ let
+  # Given a package revision (like "refs/tags/v12.0"), produce a short revision ("12.0").
+  shortRev = rev_:
+    let
+      rev = toString rev_;
+      baseRev = baseNameOf rev;
```
No reason to do this other way:
```suggestion
      baseRev = baseNameOf (toString rev);
```

AndersonTorres commented on lib/sources.nix:
```diff
@@ -254,6 +254,64 @@ let
+  # /nix/store paths.
+  #
+  # This uses a different implementation depending on the `pretty` argument:
+  #  false  -> name everything as "source"
```
To be honest I do not like dual-type attributes. This is a serious footgun in a duck-typed language.

And let's remember Nix is not stellar on the error report department:
https://discourse.nixos.org/t/rant-finding-errors-in-module-configs-after-incompatible-upgrades-is-dreadful/46375

Either is everything a string, say ` format ?  "short", "medium", "long"` or we stick to only two options (and then we can use a Boolean).

oxij replied:
But it is not checked by Nix but by the `config.nix` and NixOS typing infra, and the errors it generates are pretty nice, e.g. with `nameSourcesPrettily = "error"`:

```
$ nix-instantiate ./default.nix -A xlife
error: A definition for option `nameSourcesPrettily' is not of type `boolean or value "full" (singular enum)'. Definition values:
- In `nixpkgs.config': "error"
(use '--show-trace' to show detailed location information)
```

AndersonTorres replied:
Hum, less bad. 

Nonetheless, semantically it is a clear case for an enumeration with three possible states, instead of a weird Boolean-plus-an-extra.

AndersonTorres commented on pkgs/top-level/config.nix:
```diff
@@ -50,6 +50,39 @@ let
       default = false;
     };
 
+    nameSourcesPrettily =  mkOption {
```
```suggestion
    usePrettierSourceNames =  mkOption {
```

oxij replied:
Hm, but why? `nameSourcesPrettily` is shorter and means the same thing.

AndersonTorres replied:
It's shorter because it is more obscure.
verb-adjective-name is clearer than name-name-adverb.

On the other hand, it depends on the previous tristate vs boolean-plus-state discussion.

oxij replied:
Changed to `fetchedSourceNameDefault` in the latest version.

oxij committed:
fetchsvn: simplify `repoToName` and move it outside the thunk

This is more efficient.

oxij committed:
pkgs/top-level/config.nix: remove some unneded lines

philiptaron reviewed (commented):
A few curious questions as I read through this PR.

oxij commented on lib/sources.nix:
```diff
@@ -254,6 +254,62 @@ let
+  urlToName = url:
+    let
+      inherit (builtins) stringLength;
+      base = baseNameOf (lib.removeSuffix "/" (lib.last (lib.splitString ":" (toString url))));
```
No, the approximation of the basename, even in some really weird cases.

oxij commented on lib/sources.nix:
```diff
@@ -254,6 +254,62 @@ let
+      base = baseNameOf (lib.removeSuffix "/" (lib.last (lib.splitString ":" (toString url))));
+      # chop away one git or archive-related extension
+      removeExt = name: let
+          matchExt = builtins.match "(.*)\\.(git|tar|zip|gz|tgz|bz|tbz|bz2|tbz2|lzma|txz|xz|zstd)" name;
```
Agreed. Done.

oxij commented on lib/sources.nix:
```diff
@@ -254,6 +254,62 @@ let
       outPath = builtins.path { inherit filter name; path = origSrc; };
     };
 
+  # urlToName : (URL | Path | String) -> String
```
Well, with #49862 applied, whole of `nixpkgs` is a set of test for this, kind of.

I would not be against adding them in the future, but adding them now, when the default format is likely to change after (if ever) this gets into master and people start using it, would just ossify these specific decisions. And they are pretty random.

oxij commented on lib/sources.nix:
```diff
@@ -254,6 +254,62 @@ let
+  shortRev = rev:
+    let
+      baseRev = baseNameOf (toString rev);
+      matchHash = builtins.match "[a-f0-9]+" baseRev;
```
AFAIK, yes.

oxij commented on lib/sources.nix:
```diff
@@ -254,6 +254,62 @@ let
+      matchHash = builtins.match "[a-f0-9]+" baseRev;
+      matchVer = builtins.match "([A-Za-z]+[-_. ]?)*(v)?([0-9.]+.*)" baseRev;
+    in
+      if matchHash != null then builtins.substring 0 7 baseRev
```
No idea, this is just the convention `fetchgit` uses.

oxij commented on lib/sources.nix:
```diff
@@ -254,6 +254,62 @@ let
+  repoRevToNameFull = repo_: rev_: suffix_:
+    let
+      repo = urlToName repo_;
+      rev = if rev_ != null then "-${shortRev rev_}" else "";
```
Well, yes, but this is **a good default** which covers 99.99% of nixpkgs without any changes to specific packages needed. A packages that needs something else will just set `src.name` to something else.

oxij commented on lib/sources.nix:
```diff
@@ -254,6 +254,62 @@ let
+    let
+      baseRev = baseNameOf (toString rev);
+      matchHash = builtins.match "[a-f0-9]+" baseRev;
+      matchVer = builtins.match "([A-Za-z]+[-_. ]?)*(v)?([0-9.]+.*)" baseRev;
```
Then, `llama.cpp` will have to set `src.name` explicitly.

mweinelt commented on pkgs/build-support/fetchgitiles/default.nix:
nixfmt 'pkgs/build-support/fetchgitiles/default.nix'

oxij replied:
Indeed. Thanks!

oxij committed:
lib, treewide: introduce `repoRevToName`, use it in most `fetch*` functions

This patch adds `lib.repoRevToName` function that generalizes away most of the
code used for derivation name generation by `fetch*` functions (`fetchzip`,
`fetchFromGitHub`, etc, except those which are delayed until latter commits
for mass-rebuild reasons).

It's first argument controls how the resulting name will look (see below).

Since `lib` has no equivalent of Nixpkgs' `config`, this patch adds
`config.fetchedSourceNameDefault` option to Nixpkgs and then re-exposes
`lib.repoRevToName config.fetchedSourceNameDefault` expression as
`pkgs.repoRevToNameMaybe` which is then used in `fetch*` derivations.

The result is that different values of `config.fetchedSourceNameDefault` now
control how the `src` derivations produced by `fetch*` functions are to be
named, e.g.:

- `fetchedSourceNameDefault = "source"` (the default):

  ```
  $ nix-instantiate -A fuse.src
  /nix/store/<hash>-source.drv
  ```

- `fetchedSourceNameDefault = "versioned"`:

  ```
  $ nix-instantiate -A fuse.src
  /nix/store/<hash>-libfuse-2.9.9-source.drv
  ```

- `fetchedSourceNameDefault = "full"`:

  ```
  $ nix-instantiate -A fuse.src
  /nix/store/<hash>-libfuse-2.9.9-github-source.drv
  ```

See the documentation of `config.fetchedSourceNameDefault` for more info.

oxij commented:
ping @AndersonTorres specifically, since your review requested changes that were since resolved, IMHO.

Also, somewhat relatedly, https://github.com/NixOS/nixpkgs/pull/391322 is a follow-up to https://github.com/NixOS/nixpkgs/commit/66209649883659e071b686d348de491774f57f23 which changed the version without changing the hash, which would not have happened if @mweinelt used this PR with `fetchedSourceNameDefault = "versioned"` set. :]

Such errors, how frequent are they? :thinking:

Just saying.

AndersonTorres reviewed (approved):
I would complain about nixfmt not having its own commit. But nothing relevant is being lost.

So, LGTM.

mweinelt commented:
> Such errors, how frequent are they? 🤔

The cause is limitations in the python bulk updater. These commits are under my name, but `update-python-libraries` created them.

oxij commented:
@AndersonTorres Thanks! Also, mass-reformat now ate those `nixfmt` changes anyway.

@mweinelt Well, okay, but it would have been caught by this infra.

Also, rebased onto `master` to fix conflicts.

oxij commented:
Ping. What's holding this up?

oxij commented:
Thanks, rebase LGTM.

oxij commented:
Rebased again to fix conflicts. And it is now stuck at Darwin check again. And the previous time it was stuck right there for a month. So, I don't think it's likely this will get any better anytime soon.
